jni/src/com_amazon_opendistroforelasticsearch_knn_index_v2011_KNNIndex.cpp

```diff
@@ -38,196 +39,135 @@ using similarity::Object;
         }
-        has_exception_in_stack(env);
         return results;
     } catch(...) {
-        catch_cpp_exception_and_throw_java(env);
+        knn_jni::CatchCppExceptionAndThrowJava(env);
     }
-    return NULL;
+    return nullptr;
 }
 
+
 JNIEXPORT jlong JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_v2011_KNNIndex_init(JNIEnv* env, jclass cls,  jstring indexPath, jobjectArray algoParams, jstring spaceType)
 {
-    IndexWrapper *indexWrapper = NULL;
+    IndexWrapper *indexWrapper = nullptr;
     try {
-        const char *indexPathCStr = env->GetStringUTFChars(indexPath, 0);
-        string indexPathString(indexPathCStr);
-        env->ReleaseStringUTFChars(indexPath, indexPathCStr);
-        has_exception_in_stack(env);
-
-        // Load index from file (may throw)
-        const char *spaceTypeCStr = env->GetStringUTFChars(spaceType, 0);
-        string spaceTypeString(spaceTypeCStr);
-        env->ReleaseStringUTFChars(spaceType, spaceTypeCStr);
-        has_exception_in_stack(env);
-        IndexWrapper *indexWrapper = new IndexWrapper(spaceTypeString);
-        indexWrapper->index->LoadIndex(indexPathString);
+        string indexPathCppString = knn_jni::GetStringJenv(env, indexPath);
+        string spaceTypeCppString = knn_jni::GetStringJenv(env, spaceType);
+
+        indexWrapper = new IndexWrapper(spaceTypeCppString);
+        indexWrapper->index->LoadIndex(indexPathCppString);
 
         // Parse and set query params
-        int paramsCount = env->GetArrayLength(algoParams);
-        vector<string> paramsList;
-        for (int i=0; i<paramsCount; i++) {
-            jstring param = (jstring) (env->GetObjectArrayElement(algoParams, i));
-            const char *rawString = env->GetStringUTFChars(param, 0);
-            paramsList.push_back(rawString);
-            env->ReleaseStringUTFChars(param, rawString);
-        }
+        auto paramsList = knn_jni::GetVectorOfStrings(env, algoParams);
         indexWrapper->index->SetQueryTimeParams(AnyParams(paramsList));
-        has_exception_in_stack(env);
 
         return (jlong) indexWrapper;
     }
     // nmslib seems to throw std::runtime_error if the index cannot be read (which
     // is the only known failure mode for init()).
     catch (...) {
-        if (indexWrapper) delete indexWrapper;
-        catch_cpp_exception_and_throw_java(env);
+        delete indexWrapper;
+        knn_jni::CatchCppExceptionAndThrowJava(env);
     }
     return NULL;
 }
 
+
 JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_v2011_KNNIndex_gc(JNIEnv* env, jclass cls,  jlong indexPointer)
 {
-    try {
-        IndexWrapper *indexWrapper = reinterpret_cast<IndexWrapper*>(indexPointer);
-        has_exception_in_stack(env);
-        delete indexWrapper;
-        has_exception_in_stack(env);
-    }
-    catch (...) {
-        catch_cpp_exception_and_throw_java(env);
-    }
+    auto *indexWrapper = reinterpret_cast<IndexWrapper*>(indexPointer);
```

Comment:
Minor: Could there be possibility of exception here?

Reply:
No it will not: https://stackoverflow.com/questions/36916445/can-a-reinterpret-cast-itself-cause-an-exception